src/uvw/resource.hpp

```diff
@@ -29,19 +29,19 @@ class Resource: public Emitter<T>, public std::enable_shared_from_this<T> {
           pLoop{std::move(ref)},
           resource{}
     {
-        this->template get<U>()->data = static_cast<T*>(this);
+        resource.data = static_cast<T*>(this);
     }
 
     auto parent() const noexcept {
         return pLoop->loop.get();
     }
 
-    template<typename R>
+    template<typename R = U>
```

Comment:
I'm not too sure about that. I tried playing around with some code to help me understand how it would work. It seems `auto` deduces pretty well for the considerably new feature it is.
[Here's a test program](http://coliru.stacked-crooked.com/a/611d106f1a945420) demonstrating the same.
I'm not sure if it is compiler specific, I can try running the same program in VS on my machine.
Also, if it comes to it, shouldn't we be choosing `const_cast<const U*>`?